File: CSharpWithLanguageExt/Transfer/Domain/Events.cs
using OneOf;

using Lib.Types;

namespace Bank.Transfer.Domain;

using TransferRecipientEvent = OneOf<
   RegisteredInternalTransferRecipient,
   RegisteredDomesticTransferRecipient,
   RegisteredInternationalTransferRecipient
>;

public record TransferPending(
   Guid EntityId,
   DateTime Date,
   DateTime Timestamp,
   TransferRecipient Recipient,
   decimal DebitedAmount,
   string Reference
)
: Event(EntityId, Timestamp, nameof(TransferPending), 1.0F);

public record TransferApproved(
   Guid EntityId,
   DateTime Date,
   DateTime Timestamp,
   TransferRecipient Recipient,
   decimal DebitedAmount
)
: Event(EntityId, Timestamp, nameof(TransferApproved), 1.0F);

public record TransferRejected(
   Guid EntityId,
   DateTime Date,
   DateTime Timestamp,
   TransferRecipient Recipient,
   decimal DebitedAmount,
   string Reason
)
: Event(EntityId, Timestamp, nameof(TransferRejected), 1.0F);

public record RegisteredInternalTransferRecipient(
   Guid EntityId,
   DateTime Timestamp,
   string LastName,
   string FirstName,
   string AccountNumber
)
: Event(EntityId, Timestamp, nameof(RegisteredInternalTransferRecipient))
{
   public TransferRecipient AsTransferRecipient() =>
      new TransferRecipient(
         LastName,
         FirstName,
         AccountNumber,
         RecipientAccountEnvironment.Internal,
         RecipientAccountIdentificationStrategy.AccountID
      );
}

public record RegisteredDomesticTransferRecipient(
   Guid EntityId,
   string LastName,
   string FirstName,
   string RoutingNumber,
   string AccountNumber,
   DateTime Timestamp
)
: Event(EntityId, Timestamp, nameof(RegisteredDomesticTransferRecipient))
{
   public TransferRecipient AsTransferRecipient() =>
      new TransferRecipient(
         LastName,
         FirstName,
         AccountNumber,
         RecipientAccountEnvironment.Domestic,
         RecipientAccountIdentificationStrategy.AccountID,
         RoutingNumber
      );
}

public record RegisteredInternationalTransferRecipient(
   Guid EntityId,
   string LastName,
   string FirstName,
   string Identification,
   RecipientAccountIdentificationStrategy IdentificationStrategy,
   string Currency,
   DateTime Timestamp
)
: Event(EntityId, Timestamp, nameof(RegisteredInternationalTransferRecipient))
{
   public TransferRecipient AsTransferRecipient() =>
      new TransferRecipient(
         LastName,
         FirstName,
         Identification,
         RecipientAccountEnvironment.International,
         IdentificationStrategy,
         Currency: Currency
      );
}

public static class TransferEventsExt {
   public static Event Unwrap(this TransferRecipientEvent evtWrapped) =>
      evtWrapped.Match<Event>(
         _ => evtWrapped.AsT0,
         _ => evtWrapped.AsT1,
         _ => evtWrapped.AsT2
      );
}

public static class TransferResponseToCommand {
   public static ApproveTransferCmd Approve(TransferPending evt) {
      return new ApproveTransferCmd(
         evt.EntityId,
         evt.Recipient,
         evt.Date,
         evt.DebitedAmount
      );
   }

   public static RejectTransferCmd Reject(TransferPending evt, string reason) {
      return new RejectTransferCmd(
         evt.EntityId,
         evt.Recipient,
         evt.Date,
         evt.DebitedAmount,
         reason
      );
   }
}

File: AkkaIntegrated/Transfer/Domain/Events.fs
namespace Bank.Transfer.Domain

open System

open Lib.Types

type AckReceipt = string

type RecipientAccountEnvironment =
   | Internal
   | Domestic
   | International

type RecipientAccountIdentificationStrategy =
   | AccountId
   | SwiftBIC
   | IBAN
   | NationalID

type TransferRecipient = {
   LastName: string
   FirstName: string
   Identification: string
   AccountEnvironment: RecipientAccountEnvironment
   IdentificationStrategy: RecipientAccountIdentificationStrategy
   RoutingNumber: string option
   Currency: Currency
}

type TransferPending = {
   Recipient: TransferRecipient
   Date: DateTime
   DebitedAmount: decimal
   Reference: string option
}

type TransferApproved = {
   Recipient: TransferRecipient
   Date: DateTime
   DebitedAmount: decimal
   AckReceipt: AckReceipt option
}

type TransferRejected = {
   Recipient: TransferRecipient
   Date: DateTime
   DebitedAmount: decimal
   Reason: string
}

type RegisteredInternalTransferRecipient = {
   LastName: string
   FirstName: string
   AccountNumber: string
   Currency: Currency
   AccountEnvironment: RecipientAccountEnvironment
} with

   member x.toRecipient() = {
      FirstName = x.FirstName
      LastName = x.LastName
      AccountEnvironment = x.AccountEnvironment
      Identification = x.AccountNumber
      IdentificationStrategy = RecipientAccountIdentificationStrategy.AccountId
      Currency = x.Currency
      RoutingNumber = None
   }

type RegisteredDomesticTransferRecipient = {
   LastName: string
   FirstName: string
   RoutingNumber: string option
   AccountNumber: string
   Currency: Currency
   AccountEnvironment: RecipientAccountEnvironment
} with

   member x.toRecipient() = {
      FirstName = x.FirstName
      LastName = x.LastName
      AccountEnvironment = x.AccountEnvironment
      Identification = x.AccountNumber
      IdentificationStrategy = RecipientAccountIdentificationStrategy.AccountId
      Currency = x.Currency
      RoutingNumber = x.RoutingNumber
   }

type RegisteredInternationalTransferRecipient = {
   LastName: string
   FirstName: string
   Identification: string
   IdentificationStrategy: RecipientAccountIdentificationStrategy
   Currency: Currency
   AccountEnvironment: RecipientAccountEnvironment
} with

   member x.toRecipient() = {
      FirstName = x.FirstName
      LastName = x.LastName
      AccountEnvironment = x.AccountEnvironment
      Identification = x.Identification
      IdentificationStrategy = x.IdentificationStrategy
      Currency = x.Currency
      RoutingNumber = None
   }

type TransferDeposited = {
   DepositedAmount: decimal
   Origin: string
}
